### DISCUSSION TITLE
Any hints / tips for using the following a GPX feature?

### DISCUSSION THREAD

**Discussion by yellowbarchetta** [cagnulein/qdomyos-zwift#1893] [Q&A] [answered]
Tried it out this morning following a 12 mile route I ran a couple of weeks ago.  The gpx file is fine, but the "instructions" that the file sent to the treadmill were inconsistent, and after just over a mile it slowed things down.  During the first mile the pace wavered a lot more than the gpx file did.  
I suspect user error on my part.
Any hints or tips on how to use it properly?  And how could I see the map at the same time on the phone (android), or is this not possible?  (Need to keep the laptop free so I can watch movies etc and I don't have admin rights on it either).

**Comment by cagnulein** [comment]
are you talking about the video+gpx feature? if so share a debug log. also @Bepo7012 could be interested

**Reply by yellowbarchetta** [reply]
For some reason the debug log didn't get created (or I deleted it!) so next time I'll try the same file and see what happens.  I'm just talking about the "Open GPX" feature on the main menu, no idea about the video+gpx.  

I did also try to simulate it with the "fake treadmill" but that doesn't seem to work with the "Open GPX" feature (though again, this could be user error!)

**Reply by cagnulein** [reply]
> For some reason the debug log didn't get created (or I deleted it!) so next time I'll try the same file and see what happens.  I'm just talking about the "Open GPX" feature on the main menu, no idea about the video+gpx.  


i was talking about the ability to use qz as rouvy or kinomap to sync a video with the gpx file!


> 
> 
> 
> I did also try to simulate it with the "fake treadmill" but that doesn't seem to work with the "Open GPX" feature (though again, this could be user error!)

did you enable the  speed forcing in the treadmill settings? disable it and it will work because they are gpx for bike

**Reply by cagnulein** [reply]
for the map, yes, you can see on the same phone. There is the map icon in the top right corner.

**Reply by yellowbarchetta** [reply]
Without speed forcing the app didn't appear to control the speed of the treadmill - but I need to disable it for the fake treadmill?  OK, will try that.  Again, maybe user error - I will disable that and see what happens.

**Comment by cagnulein** [comment]
yes disable it and then the inclination at least will change following the
gpx
or at least it should :)


Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno dom 17 dic 2023 alle 22:21 yellowbarchetta <
***@***.***> ha scritto:

**Reply by yellowbarchetta** [reply]
Tried with the fake treadmill and got the same result as I did with the live one - crazy stuff happening after just after a mile which doesn't match the original GPX.  Files attached!
[drive-download-20231218T085635Z-001.zip](https://github.com/cagnulein/qdomyos-zwift/files/13701543/drive-download-20231218T085635Z-001.zip)
[22972-Activity_2023-11-12_07-05_80796612.zip](https://github.com/cagnulein/qdomyos-zwift/files/13701629/22972-Activity_2023-11-12_07-05_80796612.zip)

**Reply by cagnulein** [reply]
@yellowbarchetta define crazy stuff please :) i mean, in the maps? in the inclination? I also saw that you have the treadmill_simulate_inclination_with_speed setting set to true, this will add some crazyness to this. Did you enable it by mistake?

**Reply by yellowbarchetta** [reply]
That might be what I'm looking for to help me out.  Though the .fit file leapt from the UK to Australia (hence "crazy stuff"), but more boringly the speed dropped to 0.5mph from about 7mph for no discernible reason.

I posted it here in discussions rather than bugs as i'm sure it's something I am doing wrong or not understanding about how the feature works.  My assumption was that you simply uploaded a gpx and then the treadmill would both run at the speed that the gpx had previously recorded, and simulate inclination as the file had recorded too.

**Reply by cagnulein** [reply]
the fit file generated has the right location 

![image](https://github.com/cagnulein/qdomyos-zwift/assets/48216/5b8f2fba-e2d1-4f38-b48b-9a6b36ef1137)

are you talking about what you see on the map on QZ in realtime?

> but more boringly the speed dropped to 0.5mph from about 7mph for no discernible reason.

with the speed force disabled, you're controlling the speed. am I missing something? Sorry probably I'm not understanding the issue completely

> I posted it here in discussions rather than bugs as i'm sure it's something I am doing wrong or not understanding about how the feature works. 

sure no problem

> My assumption was that you simply uploaded a gpx and then the treadmill would both run at the speed that the gpx had previously recorded, and simulate inclination as the file had recorded too.

exactly, it's the purpose. I used it with the same scenario several times last year. If i didn't add some new bugs, it should still work.
Let me know

**Reply by yellowbarchetta** [reply]
That's not the map I get (though the first part is correct), and I can't use fitfileviewer to show maps like you can (is that a dev option?)

But if you show the rows especially for the 2nd and 3rd records you can see it leap from lat/long +53 / -2.7 to -50.4 / 167.9 and then back into the Irish Sea at 53.7/ -3.4
![image](https://github.com/cagnulein/qdomyos-zwift/assets/58736713/e97adf20-883a-4abd-b9eb-3b4fbc9a3653)

Here's the .fit file when uploaded to GPS visualiser overall showing that being mapped.
![image](https://github.com/cagnulein/qdomyos-zwift/assets/58736713/cd225573-4588-4af6-9f5d-4d44476607cb)
And showing the first part only
![image](https://github.com/cagnulein/qdomyos-zwift/assets/58736713/6883e5e9-af1f-4125-9ce8-6cf2c5d2a3b2)

That leap across the world was there visually within QZ as the accumulated distance suddenly went up to 36000 miles.

**Reply by yellowbarchetta** [reply]
I'm going to have a look and see if there is any issue in the source data of the GPX file that makes these things go crazy (e.g. dodgy altitude data or something like that) which would explain things.

So, for me to do what I thought I wanted it to do with my Horizon treadmill, am I right in thinking these are the settings?

- Treadmill speed forcing - ON (this will make the TM run at the speed that I actually did the real world run at)
- Simulate incline with speed - OFF (incline will adjust as per the file source)
- Open GPX - use this to load the run
- Tiles - don't touch any of the tiles to adjust speed / inclination as QZ will read from the gpx file to control it

**Reply by yellowbarchetta** [reply]
OK, just ran it again this time with speed forcing off and simulate incline with speed off (on Fake Treadmill).  What I can see visually happening at around the 1 mile (1609 metre) mark is that the "inclination" field suddenly starts showing extreme negative inclination (-38 I think was the biggest number).  This correlates with what I've seen in the real world treadmill where the pace dropped (when I had speed and incline linked).  When I was on the "real" treadmill I didn't see this because (I assume) the minimum inclination that could be displayed on QZ and on the TM itself was "0.0", but presumably behind the scenes the negative incline was being processed.
What's really odd is that I can't see that negative incline (not to that degree anyway!) in the source file.  
Can you replicate this?  Upload the gpx file I've shared, run the "fake treadmill" option as above and set speed to something reasonably quick so that after 5 or 6 minutes it gets to the problem section?
Debug attached for this session.
[debug-Mon_Dec_18_11_30_39_2023.log](https://github.com/cagnulein/qdomyos-zwift/files/13703341/debug-Mon_Dec_18_11_30_39_2023.log)

**Comment by cagnulein** [comment]
> That's not the map I get (though the first part is correct), and I can't use fitfileviewer to show maps like you can (is that a dev option?)

nope user mode, check the button on the top 

![image](https://github.com/cagnulein/qdomyos-zwift/assets/48216/1d0d8023-7959-4921-959c-45b6daef3766)




> But if you show the rows especially for the 2nd and 3rd records you can see it leap from lat/long +53 / -2.7 to -50.4 / 167.9 and then back into the Irish Sea at 53.7/ -3.4

i guess we are using different fit file, because i don't see this in the fit file that you sent, am I going crazy? :D



> So, for me to do what I thought I wanted it to do with my Horizon treadmill, am I right in thinking these are the settings?
> 
> * Treadmill speed forcing - ON (this will make the TM run at the speed that I actually did the real world run at)
> * Simulate incline with speed - OFF (incline will adjust as per the file source)
> * Open GPX - use this to load the run
> * Tiles - don't touch any of the tiles to adjust speed / inclination as QZ will read from the gpx file to control it

yes but there is a missing step after the open gpx: you have to press pause/start to load the training program. I mean it applies only to a new session (or just press start if the treadmill was already on pause)

I'm checking now your last log, hold on

**Reply by yellowbarchetta** [reply]
I don't see that user / developer mode switch - ah, now I've found it - I needed to run the website in incognito mode for some reason to get those flags.   Yes, the map doesn't show the "crazy" stuff, but if you go to the penultimate page of record items (timestamp around 22:27) you should see the lat / long numbers leap.  Though to be honest I've not been able to recreate that so it might have been just one odd "glitch" (I was walking round the house with my phone in my hand at the time so there's a possibility I did something to corrupt things).
I can't think that we're using different .fit files there as they have the same file name precisely - though your version of fitfileviewer.com looks a little different to mine (column order is different).  File size is identical - are you looking at the right time-stamps?
The key bit to focus on I think though is the odd processing of the gpx file in relation to elevation changes.  I've repeated that and got the same output twice so there is "something" odd in there.

**Reply by cagnulein** [reply]
ok I can see the -15% incline on the gpx now, let me check this first

**Reply by yellowbarchetta** [reply]
Ah good - I think I can see it, though it's not immediately as obvious as being "wrong" (i.e. it doesn't look as if altitude changed by say 100m or so in one go), but I can see how this would screw up the pacing if the setting linking pace to elevation was ticked.  So in that sense, there's no issue other than a word of caution when using the pace-linked-to-inclination setting?  

Maybe you could code around it by smoothing out the algorithm for that - e.g. only calculate inclination over a rolling 50m or 100m distance so that it isn't so impacted by a few sudden datapoints?  But equally, probably not needed at all if the intent was to repeat a real world activity where the pace wasn't supposed to deviate from the actual pace in the run (and inclination was just supposed to be forcing the deck to raise or lower a little bit).

**Reply by cagnulein** [reply]
ok i got the issue of the peak of the negative inclination.
Check these 2 lines for example

```
T	2023-11-12 07:07:30	53.287943000	-2.714615000	77.0	124	91.0		
T	2023-11-12 07:07:36	53.288104000	-2.714723000	75.0	121	89.0
```

there is a 2 meters of negative inclination in 20meters. That's a -10%. But of course this it's a just a little peak, so it should work the average that I added in code, I guess the error is there. Let me check this now

**Reply by cagnulein** [reply]
> Maybe you could code around it by smoothing out the algorithm for that - e.g. only calculate inclination over a rolling 50m or 100m distance so that it isn't so impacted by a few sudden datapoints? But equally, probably not needed at all if the intent was to repeat a real world activity where the pace wasn't supposed to deviate from the actual pace in the run (and inclination was just supposed to be forcing the deck to raise or lower a little bit).

it's what I'm already doing, maybe there is a bug here. Let me check ;)

**Reply by yellowbarchetta** [reply]
I think the fundamental here is I used the app incorrectly!  Even if it has highlighted a minor bug!!

**Reply by cagnulein** [reply]
got the issue, in the AVG function i wasn't put a "weight" for each sample, so i was assuming that all the samples had more or less the same distance, instead, in your gpx, there are some points in the middle with very short distance. this put crazyness in the algorithm, let me fix this, should be very simple

**Reply by cagnulein** [reply]
done, i'm testing it :)
then i will check the coordinates issue

**Reply by cagnulein** [reply, marked as answer]
@yellowbarchetta it's building https://github.com/cagnulein/qdomyos-zwift/actions/runs/7248766081 tested and now it works correctly, thanks for the feedback. now i'm focusing on the coordinates issue

**Reply by cagnulein** [reply]
the problem of the coordinates is the big power jump

23:27:59 | -50,416301 | 167,901207 | 0 | 0 | - | 52.427 | 0

let me check if i see the reason in the debug log

**Reply by cagnulein** [reply]
the issue start from here

`Sun Dec 17 22:27:59 2023 1702852079899 Debug: trainprogram.cpp void trainprogram::scheduler() positionOffset 59326.525`

**Reply by cagnulein** [reply]
it's a problem from the auto lap feature, let me check!

**Reply by yellowbarchetta** [reply]
Glad the oddities have highlighted something meaningful!

Odd that auto-lap didn't create the issue this morning when I tried it, perhaps I caught the screen and logged a manual lap which has impacted things?

GPX files with varying gaps between timestamps are not too uncommon, but I suspect different watches and different download services process things in different ways.  The GPX file I used came from Runalyze which must have generated the file itself as it will originally have been synced from Garmin to Runalyse as a .fit file.  So you're at the mercy of developers on various websites here!

**Reply by cagnulein** [reply]
> Odd that auto-lap didn't create the issue this morning when I tried it, perhaps I caught the screen and logged a manual lap which has impacted things?

i saw the issue too but only 1 time, very strange, i'm investigating it.

> GPX files with varying gaps between timestamps are not too uncommon, but I suspect different watches and different download services process things in different ways. The GPX file I used came from Runalyze which must have generated the file itself as it will originally have been synced from Garmin to Runalyse as a .fit file. So you're at the mercy of developers on various websites here!

ahahah yes. anyway thanks for the feedback, it was very important to me!

**Reply by cagnulein** [reply]
@yellowbarchetta i'm not able reproduce the auto lap issue again unfortunately. give it a try to my last build and let's see if it works ok. Let me know!

**Reply by yellowbarchetta** [reply]
As a linked query, is there a simplistic way to use the app to follow a workout that can be simply put together.
e.g. 
Time based:-
10 minutes at 10kmh, 20 minutes at 15kmh and 2.0 incline, 10 minutes at 10kmh
or
Distance based
2000m at 10kmh, 5x(800m at 15kmh, 400m at 10kmh), 200m at 10kmh

Ideally with an "auto lap" created each time a new element of the workout is done?  

Garmin Connect provides a simple interface to create this sort of thing, and I'm sure it probably already exists within QZ but I've not spotted it!  (Though I think the auto-lap would be a new item)

**Reply by cagnulein** [reply]
sure! check this article https://robertoviola.cloud/2023/11/20/exploring-the-versatility-of-qz-app-your-gateway-to-a-tailored-fitness-experience/ !

**Reply by cagnulein** [reply]
@yellowbarchetta finally I solved the mistery! the 1miles madness was about the treadmill_simulate_inclination_with_speed setting. infact, since there are high negative incline the speed calculated for it was negative. Going negative, the odometer went mad as well and then all the madness :)
So with the new build you should be fine! Let me know when you will try it

**Reply by yellowbarchetta** [reply]
Glad something good came out of that!  

Interesting looking at zwofactory; I'm guessing this is a zwift limitation as it works primarily for cyclists (runners would have to guess a FTP I think).  zwiftworkout.com also looks an option, which I'll test out, but neither of them are as easy to use as the Garmin workout feature.  That exports in a .json text file.  Have you ever looked at being able to use those?  I've loaded an example below.  Maybe some of the lack of specificity in an interval like that (I have that one set up with no defined pace or time for the intervals for example) wouldn't really work on the treadmill, whereas in the "real world" it works well as all I need to do is hit the "lap" button when needed.  My ideal would be an interval which set the pace for the TM for the various segments based on time or distance, and I do have sessions created like that too, but they tend to define a range of pace, rather than a specific one - so again, not ideal for being interpreted by QZ and sent to a treadmill.
[12-x-ntervals-.json](https://github.com/cagnulein/qdomyos-zwift/files/13730164/12-x-ntervals-.json)

**Reply by cagnulein** [reply]
@yellowbarchetta check the second part of the article, you can do what you ask with the xml file on qz! let me know if something is not clear

**Reply by yellowbarchetta** [reply]
Yes, I saw that bit; would just be nice to have little online GUI for it like those zwo creating websites to make it easier to create things.  But 100% that sort of structure would be usable, just need to test a few out.

**Comment by cagnulein** [comment]
argh the debug log is truncated, let me anyway try to check if i can see it as well

**Comment by yellowbarchetta** [comment]
No worries, that part wasn't the biggest real world issue, but if I see
that discrepancy again I'll let you know - the inclination thing was the
biggest issue.

**Reply by cagnulein** [reply]
@yellowbarchetta http://80.211.67.253:8081/training.htm?fbclid=IwAR07s1MGtk-xjNPZDj_vsE89PEUqv6D5ycffYmNxDUqax8rD0bbdMFpqDLg this one :)

i didn't update it since a lot of time, but the basic functions are there already :)

**Comment by cagnulein** [comment]
there is a web gui that i created in the past. i will bring it up again and
i will share the link

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno mer 20 dic 2023 alle 18:30 yellowbarchetta <
***@***.***> ha scritto:

**Comment by yellowbarchetta** [comment]
Probably the wrong place to ask this, but with all of the builds that you've done to address queries I've raised, how do I know if I'm running a version of the app which has all of the right bits in it? Is there something which indicates when a pull request has been blended in with other versions and developments? 
It's probably basic stuff about how GitHub works that I'm missing fully understanding to be honest!
I guess I should revert to using the the play store version at some point, but I wouldn't want to do that and lose the developments that you've made (especially the pace smoothing one).

**Comment by cagnulein** [comment]
if you are downloading the master one (as you did since i have linked you
this one the last time), you are good to go since all the patches are
already merged. you can download the nightly release from the homepage of
qz in github.

the beta on the play store is aligned, it's only doesn't have the last fix
about the inclination

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno gio 21 dic 2023 alle 17:54 yellowbarchetta <
***@***.***> ha scritto:

**Comment by yellowbarchetta** [comment]
Sorry to post on this old thread, but just wanted to check my understanding.  I use the "speed offset" at 0.975 in the app so that the pace reported in the final file better matches the treadmill (both in terms of measuring the speed of the belt and in terms of perceieved effort).  That all works fine.
But when I load a GPX file to "follow", do I need to do one of the follow:-
- leave it as is
- adjust the speed offset factor back to zero so that if the gpx file says run at 7.0mph it will tell the treadmill to run at 7.0mph
- invert the speed offset (i.e. run at 1.025) so that whatever the gpx file says will cause the treadmill to "think" it is running slightly faster than that?

Or am I overthinking this and there is no connection between the speed offset and the gpx file following?

**Reply by cagnulein** [reply]
@yellowbarchetta i guess you're simply overthinking this because QZ, from the tranining program, obtains the target speed and this is one it's a pure one and it's not affected by the speed gain or offset.

but there is a thing that we may consider: when QZ will set the target speed, example 10KM/h it will ask the treadmill to set 10KM/h, but maybe, a 10.025 should be better in your case so you will have a REAL SPEED of 10KM/h. What do you think? I already did something similar for https://github.com/cagnulein/qdomyos-zwift/discussions/2121

**Reply by yellowbarchetta** [reply]
Yes, i think that's the over-thinking I was doing.  I probably want the treadmill to be told to "run a bit faster than this".  I suppose what I wasn't sure about was that when the resulting .fit file is created does it get that data from what the treadmill has reported, because I think (I may be wrong) that when following a gpx file, the user can choose to increase / decrease the speed too if desired?

**Reply by cagnulein** [reply]
you can increase the target speed tile yes. But if you want i can handle the speed offset and the speed gain in the general scope

**Reply by yellowbarchetta** [reply]
I've not done this for a while, but I think these are the outputs from the last time I did it.  First set of splits is what the app reported in the .fit file and the second is the original splits from the source file.  I think this shows that the resulting file reported a slower speed than the original, which I think is where my original "concern" about this came from.  The gpx has made the TM properly run at the speed in the .gpx file. but then when the .fit file has received the speed data back from the treadmill, it has applied my 0.975 factor and as a result the actual run I did was slower than the outdoor one.  e.g. mile 1 - 535s vs 524s (97.9%), mile 2 518s vs 505s (97.5%) etc.  It's not entirely precise, probably due to roundings, but the difference is there.  

```
1) 1mi - 8:55(8:56/mi)
2) 1mi - 8:38(8:38/mi)
3) 1mi - 8:33(8:32/mi)
4) 1mi - 8:35(8:34/mi)
5) 1mi - 8:36(8:35/mi)
6) 1mi - 8:25(8:25/mi)
7) 1mi - 8:25(8:25/mi)
8) 0.06mi - 27(8:01/mi)
9) 0.04mi - 20(8:48/mi)
```


```
1) 1m - 8:44(8:44/m) 137/149bpm [1196b/mi] 111cal 6.88/7.24mph
2) 1m - 8:25(8:25/m) 147/154bpm [1238b/mi] 117cal 7.12/7.78mph
3) 1m - 8:17(8:17/m) 143/148bpm [1185b/mi] 106cal 7.24/7.95mph
4) 1m - 8:21(8:21/m) 145/149bpm [1211b/mi] 107cal 7.19/7.62mph
5) 1m - 8:21(8:21/m) 144/147bpm [1203b/mi] 102cal 7.18/7.58mph
6) 1m - 8:12(8:12/m) 147/152bpm [1206b/mi] 104cal 7.31/7.78mph
7) 1m - 8:11(8:11/m) 145/149bpm [1188b/mi] 96cal 7.33/7.89mph
8) 0.04m - 17(7:43/m) 144/145bpm [1112b/mi] 3cal 7.77/7.41mph
```

**Reply by yellowbarchetta** [reply]
So yes, I think the "fix" would be that if there is a speed offset, apply that to the instruction to the treadmill, and then that would fix the resulting .fit file being correct.

**Reply by cagnulein** [reply]
@yellowbarchetta Good news! I just checked and the code was already there! So it's already working on this way! Which version do you have of qz?

**Reply by yellowbarchetta** [reply]
I'm on 2.16.46

**Comment by cagnulein** [comment]
Ok so you have already this patch, just added :)

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno lun 11 mar 2024 alle 20:36 yellowbarchetta <
***@***.***> ha scritto:

**Reply by yellowbarchetta** [reply]
Great - so just so that I understand, is this saying that if there is a figure other than 1.0 in the speed offset setting, then this offset figure is applied to the instruction from the app to the treadmill without the user having to do anything else to speed things up?  I can test this out tonight maybe, set the TM to follow a gpx file, and then after 1 mile check that the app display shows me the right lap pace.

**Reply by cagnulein** [reply]
> 1.0 in the speed offset setting

other than 1.0 on the speed gain and 0.0 from the speed offset

> is applied to the instruction from the app to the treadmill without the user having to do anything else to speed things up?

exactly!

>  I can test this out tonight maybe, set the TM to follow a gpx file, and then after 1 mile check that the app display shows me the right lap pace.

oky!
